core-processor/src/ext.rs

```diff
@@ -616,11 +624,28 @@ impl EnvExt for Ext {
         Ok(())
     }
 
-    fn read(&mut self) -> Result<&[u8], Self::Error> {
+    fn read(&mut self, at: u32, len: u32) -> Result<&[u8], Self::Error> {
         let size = self
             .size()?
             .try_into()
             .unwrap_or_else(|_| unreachable!("size of the payload is a known constant: gear_core::message::MAX_PAYLOAD_SIZE < u32::MAX"));
+
+        // Verify read is correct
+        let last_idx = at
+            .checked_add(len)
+            .ok_or(ProcessorError::ReadLenOverflow(at, len))?;
+
+        {
+            let msg = self.context.message_context.current().payload();
+
+            if last_idx as usize > msg.len() {
+                return Err(ProcessorError::ReadWrongRange(
+                    at..last_idx,
+                    msg.len() as u32,
+                ));
+            }
+        }
+
         self.charge_gas_runtime(RuntimeCosts::Read(size))?;
 
         Ok(self.context.message_context.current().payload())
```

Comment:
you still return whole payload slice, isn't it?
I thought that outcome of the fn should already be &[at, at+len]

Reply:
done